backend/api/routes/files.py

```diff
@@ -0,0 +1,227 @@
+    file_id: UUID,
+    project: Project = Depends(validated_project),
+    session: Session = Depends(gen_session),
+) -> File:
+    """Depends()-able file from request, ensuring it exists"""
+    stmt = select(File).filter_by(id=file_id).filter_by(project_id=project.id)
+    file = session.execute(stmt).scalar()
+    if not file:
+        raise HTTPException(codes.NOT_FOUND, f"File not found: {file_id}")
+    return file
+
+
+def read_file_in_chunks(reader: BinaryIO, chunck_size=2048) -> Iterator[bytes]:
+    """Read Big file chunk by chunk. Default chunk size is 2k"""
+    while True:
+        chunk = reader.read(chunck_size)
+        if not chunk:
+            break
+        yield chunk
+    reader.seek(0)
+
+
+def generate_file_hash(file: BinaryIO) -> str:
+    """Generate sha256 hash of a file, optimized for large files"""
+    hasher = hashlib.sha256()
+    for chunk in read_file_in_chunks(file):
+        hasher.update(chunk)
+    return hasher.hexdigest()
+
+
+def upload_file(location: Path, file: BinaryIO) -> Path:
+    """Saves a binary file to a specific location and returns its path."""
+    if not location.exists():
+        os.makedirs(location, exist_ok=True)
+    file_location = Path(tempfile.NamedTemporaryFile(dir=location, delete=False).name)
+    with open(file_location, "wb") as file_object:
+        for chunk in read_file_in_chunks(file):
+            file_object.write(chunk)
+    return file_location
+
+
+def validate_uploaded_file(upload_file: UploadFile):
+    """
+    Validates the uploaded file to ensure it meets the requirements.
+
+    Args:
+        upload_file (UploadFile): The uploaded file object.
+
+    Returns:
+        tuple: A tuple containing filename, size, and mimetype of the file.
+
+    Raises:
+        HTTPException: If the filename is invalid, the file is empty.
+    """
+    filename = upload_file.filename
+    size = upload_file.size
+
+    if not filename:
+        raise HTTPException(
+            status_code=codes.NOT_ACCEPTABLE, detail="Filename is invalid."
+        )
+
+    if not size or size == 0:
+        raise HTTPException(status_code=codes.NOT_ACCEPTABLE, detail="Emtpy file.")
+
+    mimetype = filesystem.get_content_mimetype(upload_file.file.read(2048))
+    upload_file.file.seek(0)
+
+    return (filename, size, mimetype)
+
+
+def validate_project_quota(file_size: int, project: Project):
+    """Validates total size of uploaded files to ensure it meets the requirements."""
+    total_size = sum([file.filesize for file in project.files]) + file_size
+    if total_size > BackendConf.maximum_project_quota:
+        raise HTTPException(
+            status_code=codes.REQUEST_ENTITY_TOO_LARGE,
+            detail="Uploaded files exceeded quota",
+        )
+
+
+@router.post("/{project_id}/files", status_code=codes.CREATED)
+async def create_file(
+    uploaded_file: UploadFile,
+    project: Project = Depends(validated_project),
+    session: Session = Depends(gen_session),
+) -> FileModel:
```

Comment:
The api can only upload one file at a time, making it easier to track the progress of file uploads.